From a977b3561800ac85309365707d2bf0d35bb25117 Mon Sep 17 00:00:00 2001
From: Loc Vu <loc.vu.zn@renesas.com>
Date: Fri, 14 Jul 2023 16:10:24 +0700
Subject: [PATCH 3/3] rzg3s: Add SMC command for USB PWRRDY terminal control

Add SMC command to support USB_PWRRDY terminal control. Example usage from
Linux kernel as following:

a. Set SYS_USB_PWRRDY=0 when supply power for USB
   arm_smccc_smc(RZ_SIP_SVC_SET_PCIE_RST_RSMB, 0xd70, 0x0, 0, 0, 0, 0, 0, &local_res);
b. Set SYS_USB_PWRRDY=1 when stop supplying power for USB
   arm_smccc_smc(RZ_SIP_SVC_SET_PCIE_RST_RSMB, 0xd70, 0x1, 0, 0, 0, 0, 0, &local_res);

Signed-off-by: Loc Vu <loc.vu.zn@renesas.com>
---
 plat/renesas/rz/common/include/rz_sip_svc.h  |  3 +++
 plat/renesas/rz/common/rz_plat_sip_handler.c | 13 +++++++++++++
 plat/renesas/rz/soc/g3s/include/rz_soc_def.h |  1 +
 3 files changed, 17 insertions(+)

diff --git a/plat/renesas/rz/common/include/rz_sip_svc.h b/plat/renesas/rz/common/include/rz_sip_svc.h
index 542e33a1f..7a89c4cbb 100644
--- a/plat/renesas/rz/common/include/rz_sip_svc.h
+++ b/plat/renesas/rz/common/include/rz_sip_svc.h
@@ -16,4 +16,7 @@
 /* Function ID to set PCIE RST_RSM_B */
 #define RZ_SIP_SVC_SET_PCIE_RST_RSMB		U(0x82000013)
 
+/* Function ID to set USB Power Ready. */
+#define RZ_SIP_SVC_SET_USB_PWRRDY		U(0x82000014)
+
 #endif /* __RZ_SIP_SVC_H__ */
diff --git a/plat/renesas/rz/common/rz_plat_sip_handler.c b/plat/renesas/rz/common/rz_plat_sip_handler.c
index 9d2e976b3..334b5e75d 100644
--- a/plat/renesas/rz/common/rz_plat_sip_handler.c
+++ b/plat/renesas/rz/common/rz_plat_sip_handler.c
@@ -59,6 +59,17 @@ static uintptr_t rz_otp_handler_set_pcie(void *handle, u_register_t x1, u_regist
 	}
 }
 
+static uintptr_t rz_otp_handler_set_usb(void *handle, u_register_t x1, u_register_t x2)
+{
+	if ((uintptr_t)x1 == RZG3S_SYSC_USB_PWRRDY_OFFSET) {
+		mmio_write_32(RZG3S_SYSC_BASE + (uintptr_t)x1, (uint32_t)x2);
+		SMC_RET0(handle);
+	} else {
+		WARN("%s: Invalid SYSC_USB_PWRRDY address\n", __func__);
+		SMC_RET1(handle, SMC_ARCH_CALL_INVAL_PARAM);
+	}
+}
+
 uintptr_t rz_plat_sip_handler(uint32_t smc_fid,
 					u_register_t x1,
 					u_register_t x2,
@@ -75,6 +86,8 @@ uintptr_t rz_plat_sip_handler(uint32_t smc_fid,
 		return rz_otp_handler_chipid(handle, x1, flags);
 	case RZ_SIP_SVC_SET_PCIE_RST_RSMB:
 		return rz_otp_handler_set_pcie(handle, x1, x2);
+	case RZ_SIP_SVC_SET_USB_PWRRDY:
+		return rz_otp_handler_set_usb(handle, x1, x2);
 	default:
 		WARN("%s: Unimplemented RZ SiP Service Call: 0x%x\n", __func__, smc_fid);
 		SMC_RET1(handle, SMC_UNK);
diff --git a/plat/renesas/rz/soc/g3s/include/rz_soc_def.h b/plat/renesas/rz/soc/g3s/include/rz_soc_def.h
index d49e3771b..365602da2 100755
--- a/plat/renesas/rz/soc/g3s/include/rz_soc_def.h
+++ b/plat/renesas/rz/soc/g3s/include/rz_soc_def.h
@@ -123,6 +123,7 @@
 
 #define RZ_SOC_SYSC_BASE_DEVID			SYS_LSI_DEVID
 #define RZ_SOC_OTP_BASE_CHIPID			(RZG3S_OTP_BASE + 0x1140)
+#define RZG3S_SYSC_USB_PWRRDY_OFFSET			UL(0x0D70)
 #define RZG3S_SYSC_PCIE_RST_RSM_B_OFFSET		UL(0x0D74)
 
 #define RZ_SOC_I2C_BASE					RZG3S_I2C_1_BASE
-- 
2.25.1

